compiler/dyno/lib/resolution/resolution-queries.cpp

```diff
@@ -2302,6 +2427,173 @@ CallResolutionResult resolveGeneratedCall(Context* context,
+        auto receiverQualType = tfs->formalType(0);
+
+        // receiver type matches, return true
+        if (receiverQualType.type() == type) {
+          return true;
+        }
+      }
+    }
+  }
+
+  return false;
+}
+
+bool
+needCompilerGeneratedMethod(Context* context, const Type* type,
+                            UniqueString name) {
+  if (isNameOfCompilerGeneratedMethod(name)) {
+    if (!areOverloadsPresentInDefiningScope(context, type, name)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+// TODO: should we grab the parent instantiation here? Doesn't make sense
+// to generate this for partial instantiations.
+static TypedFnSignature*
+generateInitSignature(Context* context, const CompositeType* compType) {
+  std::vector<UntypedFnSignature::FormalDetail> ufsFormals;
+  std::vector<QualifiedType> formalTypes;
+
+  // start by adding a formal for the receiver
+  auto ufsReceiver = UntypedFnSignature::FormalDetail(USTR("this"),
+                                                      false,
+                                                      nullptr);
+  ufsFormals.push_back(std::move(ufsReceiver));
+  formalTypes.push_back(QualifiedType(QualifiedType::REF, compType));
+
+  // consult the fields to build up the remaining untyped formals
+  const bool useGenericDefaults = false;
+  auto& rf = fieldsForTypeDecl(context, compType, useGenericDefaults);
+
+  // push all fields -> formals in order
+  for (int i = 0; i < rf.numFields(); i++) {
+    auto qualType = rf.fieldType(i);
+    auto name = rf.fieldName(i);
+    bool hasDefault = rf.fieldHasDefaultValue(i);
+    const uast::Decl* node = nullptr;
+    auto fd = UntypedFnSignature::FormalDetail(name, hasDefault, node);
+    ufsFormals.push_back(std::move(fd));
+
+    // for types & param, use the field kind, for values use 'in' intent
+    if (qualType.isType() || qualType.isParam()) {
+      formalTypes.push_back(qualType);
+    } else {
+      auto qt = QualifiedType(QualifiedType::IN, qualType.type());
+      formalTypes.push_back(std::move(qt));
+    }
+  }
+
+  // build the untyped signature
+  auto ufs = UntypedFnSignature::get(context,
+                              /*id*/ compType->id(),
+                              /*name*/ USTR("init"),
+                              /*isMethod*/ true,
+                              /*idIsFunction*/ false,
+                              /*idIsClass*/ true,
+                              /*idIsTypeConstructor*/ false,
+                              /*kind*/ uast::Function::Kind::PROC,
+                              /*formals*/ std::move(ufsFormals),
+                              /*whereClause*/ nullptr);
+
+  // now build the other pieces of the typed signature
+  auto whereClauseResult = TypedFnSignature::WHERE_NONE;
+  bool needsInstantiation = rf.isGeneric();
+  const TypedFnSignature* instantiatedFrom = nullptr;
+  const TypedFnSignature* parentFn = nullptr;
+
+  // TODO: not sure when this is non-empty
```

Comment:
I don't think this will ever need to be non-empty here (because this generates the "initial" signature for the compiler-generated init rather than an instantiation)